src/main/java/com/example/auctionapp/Util/Search/QueryBuilder.java

```diff
@@ -0,0 +1,122 @@
+package com.example.auctionapp.Util.Search;
+
+import com.example.auctionapp.enumeration.ColorEnum;
+import com.example.auctionapp.enumeration.SizeEnum;
+import com.example.auctionapp.model.Product;
+
+import javax.persistence.EntityManager;
+import javax.persistence.TypedQuery;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Order;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.ArrayList;
+import java.util.List;
+
+public class QueryBuilder {
+
+    private static final String NAME = "name";
+    private static final String SUBCATEGORY = "subcategory";
+    private static final String ID = "id";
+    private static final String COLOR = "color";
+    private static final String SIZE = "size";
+    private static final String PRICE = "price";
+    private static final String DATE_CREATED = "dateCreated";
+    private static final String END_DATE = "auctionEndDate";
+
+    private final EntityManager entityManager;
+
+    private final CriteriaBuilder cb;
+
+    private final CriteriaQuery<Product> query;
+
+    private final Root<Product> resource;
+
+    private List<Predicate> predicates;
+
+    private Order order;
+
+    private Integer pageNumber;
+
+    private Integer pageSize;
+
+    public QueryBuilder(EntityManager entityManager) {
+        this.entityManager = entityManager;
+        this.cb = entityManager.getCriteriaBuilder();
+        this.query = this.cb.createQuery(Product.class);
+        this.resource = this.query.from(Product.class);
+        this.predicates = new ArrayList<>();
+    }
+
```

Comment:
It would be good to make this class an abstract class, so that we can extract this and common logic/properties to the parent class. 
This would make this builder reusable for different types of search queries, no just for product.class